### ISSUE TITLE
GCC 9 can't compile arch/powerpc/crypto

### ISSUE THREAD

**Issue by sathnaga** [linuxppc/issues#222]
linuxppc merge https://github.com/linuxppc/linux/commit/dbaaa7d30d9553ab07742764699f6d3e03dfdc9d branch fails to compile 

Config tried: `ppc64le_defconfig`, `pseries_defconfig`
```
# make
  CALL    scripts/checksyscalls.sh
  CHK     include/generated/compile.h
  CC [M]  arch/powerpc/crypto/md5-glue.o
In file included from arch/powerpc/crypto/md5-glue.c:17:
./include/linux/module.h:132:6: error: ‘init_module’ specifies less restrictive attribute than its target ‘ppc_md5_mod_init’: ‘cold’ [-Werror=missing-attributes]
  132 |  int init_module(void) __attribute__((alias(#initfn)));
      |      ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:157:1: note: in expansion of macro ‘module_init’
  157 | module_init(ppc_md5_mod_init);
      | ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:147:19: note: ‘init_module’ target declared here
  147 | static int __init ppc_md5_mod_init(void)
      |                   ^~~~~~~~~~~~~~~~
In file included from arch/powerpc/crypto/md5-glue.c:17:
./include/linux/module.h:138:7: error: ‘cleanup_module’ specifies less restrictive attribute than its target ‘ppc_md5_mod_fini’: ‘cold’ [-Werror=missing-attributes]
  138 |  void cleanup_module(void) __attribute__((alias(#exitfn)));
      |       ^~~~~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:158:1: note: in expansion of macro ‘module_exit’
  158 | module_exit(ppc_md5_mod_fini);
      | ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:152:20: note: ‘cleanup_module’ target declared here
  152 | static void __exit ppc_md5_mod_fini(void)
      |                    ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [scripts/Makefile.build:277: arch/powerpc/crypto/md5-glue.o] Error 1
make[1]: *** [scripts/Makefile.build:492: arch/powerpc/crypto] Error 2
make: *** [Makefile:1043: arch/powerpc] Error 2```

**Comment by chleroy** [comment]
No problem found on my side (Using GCC 8.1.0 from https://kernel.org/pub/tools/crosstool/)

[root@po16846vm linux-powerpc]# git show
commit dbaaa7d30d9553ab07742764699f6d3e03dfdc9d (HEAD -> merge, origin/merge)
Merge: bad81e52a8d3 30bac164aca7
Author: Michael Ellerman <mpe@ellerman.id.au>
Date:   Thu Jan 24 14:32:35 2019 +1100

    Automatic merge of branches 'master', 'next' and 'fixes' into merge

[root@po16846vm linux-powerpc]# make CROSS_COMPILE=powerpc64-linux- ARCH=powerpc pseries_defconfig
#
# configuration written to .config
#
[root@po16846vm linux-powerpc]# make CROSS_COMPILE=powerpc64-linux- ARCH=powerpc modules
scripts/kconfig/conf  --syncconfig Kconfig
  UPD     include/config/kernel.release
  UPD     include/generated/utsrelease.h
  CC      scripts/mod/empty.o
  MKELF   scripts/mod/elfconfig.h
  HOSTCC  scripts/mod/modpost.o
  CC      scripts/mod/devicetable-offsets.s
  UPD     scripts/mod/devicetable-offsets.h
  HOSTCC  scripts/mod/file2alias.o
  HOSTCC  scripts/mod/sumversion.o
  HOSTLD  scripts/mod/modpost
  CC      kernel/bounds.s
  UPD     include/generated/bounds.h
  CC      arch/powerpc/kernel/asm-offsets.s
  UPD     include/generated/asm-offsets.h
  CALL    scripts/checksyscalls.sh
  AS [M]  arch/powerpc/crypto/md5-asm.o
  CC [M]  arch/powerpc/crypto/md5-glue.o
  LD [M]  arch/powerpc/crypto/md5-ppc.o
  AS [M]  arch/powerpc/crypto/sha1-powerpc-asm.o
  CC [M]  arch/powerpc/crypto/sha1.o
  LD [M]  arch/powerpc/crypto/sha1-powerpc.o
  AS [M]  arch/powerpc/crypto/crc32c-vpmsum_asm.o
  CC [M]  arch/powerpc/crypto/crc32c-vpmsum_glue.o
  LD [M]  arch/powerpc/crypto/crc32c-vpmsum.o

**Comment by mpe** [comment]
Kisskb is not seeing it either: http://kisskb.ellerman.id.au/kisskb/head/3cc5258bf4fef4498d945abfddc06369d4c4b54a/

**Comment by sathnaga** [comment]
gcc version in the system:

```
# gcc --version
gcc (GCC) 9.0.0 20190119 (Red Hat 9.0.0-0.3)
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by sathnaga** [comment]
am still seeing the error with latest commit as well

# git show
commit 0194cc2cdf74db793a89f01b9cb68a22e6eee42d (HEAD -> merge, origin/merge)
Merge: 3cc5258bf4fe 12491ed354d2
Author: Michael Ellerman <mpe@ellerman.id.au>
Date:   Sun Feb 3 19:56:04 2019 +1100

    Automatic merge of branches 'master', 'next' and 'fixes' into merge


  CC [M]  arch/powerpc/crypto/md5-glue.o
In file included from arch/powerpc/crypto/md5-glue.c:17:
./include/linux/module.h:132:6: error: ‘init_module’ specifies less restrictive attribute than its target ‘ppc_md5_mod_init’: ‘cold’ [-Werror=missing-attributes]
  132 |  int init_module(void) __attribute__((alias(#initfn)));
      |      ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:157:1: note: in expansion of macro ‘module_init’
  157 | module_init(ppc_md5_mod_init);
      | ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:147:19: note: ‘init_module’ target declared here
  147 | static int __init ppc_md5_mod_init(void)
      |                   ^~~~~~~~~~~~~~~~
In file included from arch/powerpc/crypto/md5-glue.c:17:
./include/linux/module.h:138:7: error: ‘cleanup_module’ specifies less restrictive attribute than its target ‘ppc_md5_mod_fini’: ‘cold’ [-Werror=missing-attributes]
  138 |  void cleanup_module(void) __attribute__((alias(#exitfn)));
      |       ^~~~~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:158:1: note: in expansion of macro ‘module_exit’
  158 | module_exit(ppc_md5_mod_fini);
      | ^~~~~~~~~~~
arch/powerpc/crypto/md5-glue.c:152:20: note: ‘cleanup_module’ target declared here
  152 | static void __exit ppc_md5_mod_fini(void)
      |                    ^~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [scripts/Makefile.build:277: arch/powerpc/crypto/md5-glue.o] Error 1
make[1]: *** [scripts/Makefile.build:492: arch/powerpc/crypto] Error 2
make: *** [Makefile:1043: arch/powerpc] Error 2

**Comment by mpe** [comment]
Fix being discussed: https://lore.kernel.org/lkml/20190123173707.GA16603@gmail.com/

**Comment by chleroy** [comment]
Maybe look at https://github.com/zfsonlinux/zfs/issues/8329

**Comment by sathnaga** [comment]
> Fix being discussed: https://lore.kernel.org/lkml/20190123173707.GA16603@gmail.com/

@mpe any updates on this, when will it get merged in master/merge?
with this patch am able to compile without error in gcc 9.

**Comment by sathnaga** [comment]
@mpe Just checked, current linuxppc merge https://github.com/linuxppc/linux/commit/e50585e893bad2c9285e85993894b87f24ddc38f branch compiles fine with gcc 9, this issue can be closed...

**Comment by mpe** [comment]
Fixed in torvalds/linux@a6e60d84989f `include/linux/module.h: copy __init/__exit attrs to init/cleanup_module`
